File: KEX/KVG/KVGElement.m
//
//  KVGElement.m
//  KEX
//
//  Created by Daniel Schlaug on 3/1/14.
//  Copyright (c) 2014 Daniel Schlaug. All rights reserved.
//

#import "KVGElement.h"

@interface KVGElement ()

@property (nonatomic, readwrite) unichar mostSimilarUnicode;
@property (nonatomic, readwrite) unichar semanticUnicode;
@property (nonatomic, readwrite) KVGPosition position;
@property (nonatomic, readwrite) KVGRadical radical;
@property (nonatomic, readwrite) bool variant;
@property (nonatomic, readwrite) bool partial;
@property (nonatomic, readwrite) NSInteger number;
@property (strong, nonatomic, readwrite) NSArray *childElements;
@property (strong, nonatomic, readwrite) KVGElement *parentElement;
@property (strong, nonatomic, readwrite) NSArray *rootStrokes;

@end

@implementation KVGElement

#pragma mark - Properties

- (NSArray *)childElements
{
    if (!_childElements) {
        _childElements = [[NSMutableArray alloc] init];
    }
    return _childElements;
}

- (NSArray *)rootStrokes
{
    if (!_rootStrokes) {
        _rootStrokes = [[NSMutableArray alloc] init];
    }
    return _rootStrokes;
}

#pragma mark - Stroke stuff

- (NSArray *)strokes
{
    NSMutableArray *strokes = [[NSMutableArray alloc] init];
    for (KVGStroke *stroke in self.rootStrokes) {
        [strokes addObject:stroke];
    }
    for (KVGElement *element in self.childElements) {
        [strokes addObjectsFromArray:[element strokes]];
    }

    return [strokes sortedArrayUsingComparator:^NSComparisonResult(KVGStroke *obj1, KVGStroke *obj2) {
        if (obj1.strokeOrder > obj2.strokeOrder) {
            return (NSComparisonResult)NSOrderedDescending;
        }
        
        if (obj1.strokeOrder < obj2.strokeOrder) {
            return (NSComparisonResult)NSOrderedAscending;
        }
        return (NSComparisonResult)NSOrderedSame;
    }];
}

#pragma mark - Initialization

+ (KVGElement *)elementFromRXML:(RXMLElement *)elementNode
{
    KVGElement *element = [[KVGElement alloc] init];
    
    [element readAttributesFrom:elementNode];
        
    // Read the children
    [elementNode iterate:@"*" usingBlock:^(RXMLElement *elementChildNode){
        if ([elementChildNode.tag isEqualToString:@"g"])
        {
            KVGElement *newElement = [KVGElement elementFromRXML:elementChildNode];
            [(NSMutableArray *)element.childElements addObject:newElement];
            
        } else if ([elementChildNode.tag isEqualToString:@"path"]) {
            KVGStroke *stroke = [[KVGStroke alloc] initFromRXML:elementChildNode];
            [(NSMutableArray *)element.rootStrokes addObject:stroke];
            
        }
        
    }];
    
    
    return element;
}

#pragma mark - Private methods

- (void)readAttributesFrom:(RXMLElement *)XMLRepresentation {
    self.mostSimilarUnicode = [[XMLRepresentation attribute:@"element"] characterAtIndex:0];
    self.semanticUnicode = [[XMLRepresentation attribute:@"original"] characterAtIndex:0];
    
    NSString *position = [XMLRepresentation attribute:@"position"];
    if (!position) {self.position = KVGPositionNone;}
    else if ([position isEqualToString:@"left,"])   {self.position = KVGPositionLeft;}
    else if ([position isEqualToString:@"right,"])  {self.position = KVGPositionRight;}
    else if ([position isEqualToString:@"top,"])    {self.position = KVGPositionTop;}
    else if ([position isEqualToString:@"bottom,"]) {self.position = KVGPositionBottom;}
    else if ([position isEqualToString:@"nyo,"])    {self.position = KVGPositionNyo;}
    else if ([position isEqualToString:@"tare,"])   {self.position = KVGPositionTare;}
    else if ([position isEqualToString:@"kamae,"])  {self.position = KVGPositionKamae;}
    else if ([position isEqualToString:@"kamae1,"]) {self.position = KVGPositionKamae1;}
    else if ([position isEqualToString:@"kamae2"])  {self.position = KVGPositionKamae2;}
    
    NSString *radical = [XMLRepresentation attribute:@"radical"];
    if (!radical) {self.radical = KVGPositionNone;}
    else if ([radical isEqualToString:@"general"])  {self.radical = KVGRadicalGeneral;}
    else if ([radical isEqualToString:@"nelson"])   {self.radical = KVGRadicalNelson;}
    else if ([radical isEqualToString:@"tradit"])   {self.radical = KVGRadicalTradit;}
    
    NSString *variant = [XMLRepresentation attribute:@"variant"];
    if ([variant isEqualToString:@"true"]) {self.variant = true;}
    else {self.variant = false;}
    
    NSString *partial = [XMLRepresentation attribute:@"partial"];
    if ([partial isEqualToString:@"true"]) {self.partial = true;}
    else {self.partial = false;}
    
    self.number = [XMLRepresentation attributeAsInt:@"number"];
}

#pragma mark - Overrides

- (NSString *)description
{
    return [NSString stringWithFormat:
            @"KVGElement: %hu\nVariant:%@\nNumber:%ld", self.mostSimilarUnicode, self.variant ? @"true" : @"false", (long)self.number];
}

@end

File: KEX/KVG/KVGElement.swift
//
//  KVGElement.swift
//  KEX
//
//  Created by Daniel Schlaug on 6/25/14.
//  Copyright (c) 2014 Daniel Schlaug. All rights reserved.
//

import Foundation
import DanielsKit

enum KVGRadical: String {
    case General = "general"
    case Nelson = "nelson"
    case Tradit = "tradit"
}

enum KVGPosition: String {
    case Left = "left"
    case Right = "right"
    case Top = "top"
    case Bottom = "bottom"
    case Nyo = "nyo"
    case Tare = "tare"
    case Kamae = "kamae"
    case Kamae1 = "kamae1"
    case Kamae2 = "kamae2"
}

class KVGElement: NSObject {
    let mostSimilarUnicode: Character?
    let semanticUnicode: Character?
    let position: KVGPosition?
    let radical: KVGRadical?
    let variant: Bool
    let partial: Bool
    let number: Integer?
    let childElements: KVGElement[]
    let parentElement: KVGElement?
    let rootStrokes: KVGStroke[]
    
    var strokes: KVGStroke[] {
        var strokes = rootStrokes
        for childElement in childElements {
            strokes += childElement.strokes
        }
            
        strokes = sort(strokes) {$0.strokeOrder > $1.strokeOrder}
        
        return strokes
    }

    init(RXML elementNode: RXMLElement) {
        if let elementAttribute = elementNode.attribute("element"){
            mostSimilarUnicode = elementAttribute[0]
        }
        if let originalAttribute = elementNode.attribute("original") {
            semanticUnicode = originalAttribute[0]
        }
        if let positionAttribute = elementNode.attribute("position") {
            position = KVGPosition.fromRaw(positionAttribute)
        }
        if let radicalAttribute = elementNode.attribute("radical") {
            radical = KVGRadical.fromRaw(radicalAttribute)
        }
        if let variantAttribute = elementNode.attribute("variant") {
            variant = variantAttribute == "true"
        } else {variant = false}
        if let partialAttribute = elementNode.attribute("partial") {
            partial = partialAttribute == "true"
        } else {partial = false}
        if let numberAttribute = elementNode.attribute("number") {
            number = numberAttribute.toInt()
        }
        
        var childElements: KVGElement[] = []
        var rootStrokes: KVGStroke[] = []
        
        elementNode.iterate("*") {
            (elementChildNode: RXMLElement!) in
            if elementChildNode.tag == "g" {
                childElements += KVGElement(RXML: elementChildNode)
            } else if elementChildNode.tag == "path" {
                rootStrokes += KVGStroke(RXML: elementChildNode)
            }
        }
        
        self.childElements = childElements
        self.rootStrokes = rootStrokes
    }
    
}
